docs/architecture/ADR-004-interest-and-reserves.md

```diff
@@ -0,0 +1,257 @@
+// k is the x/leverage keeper and ctx contains information on the current block
+if ctx.BlockHeight() >= k.GetLastInterestBlock() + k.GetBorrowInterestEpoch() {
+    // unix times (int64 values, measured in seconds)
+    secondsElapsed := ctx.BlockTime.Unix() - k.GetLastInterestTime()
+    // accrue interest to all borrow positions at once
+    err = k.AccrueAllInterest(secondsElapsed)
+    // Store the current block's height and time in keeper
+    k.SetLastInterestBlock(ctx.BlockHeight())
+    k.SetLastInterestTime(ctx.BlockTime())
+}
+```
+
+To support the function above, the `x/leverage` module keeper must store the last unix time and block height at which interest accrued. Both values will be marshaled to binary from `sdk.Int`. Single byte prefixes can be used:
+
+```go
+key = interestLastHeightPrefix // e.g. 0x06
+key = interestLastTimePrefix // e.g. 0x07
+```
+
+### Dynamic Borrow Interest Rates
+
+Borrow interest rates are dynamic. They are calculated using the lending pool's current borrow utilization for each asset type, as well as multiple governance parameters that are decided on a per-token basis. Even the model used for the interest rate (which determines what other governance parameters are required) must itself be governed.
+
+Here are some example initial governance parameters which would be stored for a single asset type:
+```go
+InterestRateModel = 0x01 // enumeration. Which model (if multiple available) to use?
+// Model 0x01, based on Compound's JumpRateModelV2, requires the following per token:
+BaseAPY = sdk.NewDec("0.02")
+KinkAPY = sdk.NewDec("0.2")
+MaxAPY = sdk.NewDec("1.0")
+KinkUtilization = sdk.NewDec("0.8")
+```
+
+Each parameter, being stored per token, would need a keeper prefix:
+
+```go
+// For the interest rate model enumeration, which always exists for each token
+interestModelPrefix | denom | 0x00
+// For other parameters, which only exist when the interest model is a certain value
+interestParamPrefix | lengthPrefixed(denom) | paramNumber
+// example paramNumbers
+//  BaseInterest: 0x00
+//  KinkInterest: 0x01
+//  MaxInterest: 0x02
+//  KinkUtilization: 0x03
+```
+
+The `0x01` interest model shown above, based on [Compound's JumpRateModelV2 model](https://compound.finance/governance/proposals/20) defined in [this contract](https://etherscan.io/address/0xfb564da37b41b2f6b6edcc3e56fbf523bd9f2012#code), can be summarized as follows:
+
+> The (x,y) = (utilization, interest rate) graph is a line with a kink in it, defined by three points
+> At 0% utilization, there is a base interest rate `BaseInterest`
+> The kink at `KinkUtilization` utilization has interest rate `KinkInterest`
+> At 100% utilization, the interest rate is `MaxInterest`
+
+Because parameters can cease to be necessary when the interest model changes (e.g. if an interest model 0x02 only required two parameters instead of four), the governance process should delete unused parameters from the keeper when changing models if deemed necessary.
```

Comment:
I think this can become fairly complex. What concrete models do we want to support? Instead, should we rather attempt software upgrades when desiring to change the interest rate model?